### ISSUE TITLE
Backend search insights data points seem to be wrong

### ISSUE THREAD

**Issue by felixfbecker** [sourcegraph/sourcegraph#19537]
This is the current backend-based search demo insight we have on sourcegraph.com:

![image](https://user-images.githubusercontent.com/10532611/112881384-e504f400-90cb-11eb-9414-2bf962f1e97f.png)

This seems... wrong? 

- The data points are very oddly spaced (no consistent interval) and very far apart
- The usage of `fmt.Errorf()` went from the thousands, slowly down to zero (!), back up to even higher thousands.
  There is no way our use of `fmt.Errorf()` would fluctuate that much, right?!
  Maybe as a delta from one day to the next I could see why it's sometimes zero, but then again I don't think we'd add thousands of `fmt.Errorf()` in less than a month... And the chart is supposed to show counts on each day, not a delta.

cc @slimsag just in case you have any immediate thoughts.

**Comment by slimsag** [comment]
I agree, and yuck - sorry this is happening. This is definitely due to the data backfiller not working properly. The data points recorded _on and after_ the day the insight was created, will be correct.

I see a similarly concerning issue on k8s.sgdev.org - these shouldn't be zero..

<img width="784" alt="image" src="https://user-images.githubusercontent.com/3173176/112913485-8bda9780-90ae-11eb-9cb4-9a7d59427298.png">

**Comment by slimsag** [comment]
Here's what I think is happening:

1. All the data points _are correct_.
2. The data is "still being backfilled" (it is VERY slow 😞)
3. There is a bug somewhere in how it is being populated, which means older data points are being populated before newer data points - or points in the middle are being populated first - which makes the chart SUPER confusing.

It should be possible to verify if this is the case by running this locally, run:

```
DISABLE_CODE_INSIGHTS_HISTORICAL=false ./enterprise/dev/start.sh
```

Then add a code insight ([like this](https://docs.sourcegraph.com/dev/background-information/insights/backend#1-user-defines-insight-in-settings)) and see how it populates the data: it should populate it fro most-recent to 6mo ago.

When it gets part-way through (e.g. 2-3mo of data backfilled) add a new insight. Does it populate in the same logical order, or some other way?

If that doesn't sus out what is causing it, some other things we can try:

* Maybe restarting repo-updater half-way through causes some non-deterministic ordering of backfilling.
* We can check if Sourcegraph.com DB has a lot of jobs queued (hints my suspicion is right) and run the same query ([this](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/internal/insights/background/queryrunner/worker.go#L120-136) with [this order clause](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/internal/insights/background/queryrunner/worker.go#L73)) to see if the order they would be popped off the queue is right or not (based on the search query `before:` `after:` times)
* We can check the repo-updater logs on k8s.sgdev.org - something odd is going on there which causes those search queries to record zero values (not just fail) - which is concerning and may or may not be related.

I will see if I can test some of these theories and report back. Sorry the backfiller sucks :/ We can have customers disable it with `DISABLE_CODE_INSIGHTS_HISTORICAL=true` on the `repo-updater` container, if needed (but obv. that sucks for other reasons so really just need a proper fix here ASAP.)

**Comment by slimsag** [comment]
OK, so now I'm seeing the same thing on Sourcegraph.com - all the data has been wiped out:

<img width="405" alt="image" src="https://user-images.githubusercontent.com/3173176/113365196-592cdb00-930a-11eb-9227-1270fd93231a.png">

The good news here is that this might explain why the historical backfiller is taking so much longer than I expected: all the data is getting wiped out occasionally for some reason!

The bad news is.. why on earth is this happening? Maybe we still have an issue like https://github.com/sourcegraph/deploy-sourcegraph/pull/1936 ? Looking into it

**Comment by slimsag** [comment]
Confirmed the data returned by the GraphQL API does indeed look like that via [this query](https://sourcegraph.com/api/console#%7B%22query%22%3A%22%7B%5Cn%20%20insights%20%7B%5Cn%20%20%20%20nodes%20%7B%5Cn%20%20%20%20%20%20title%5Cn%20%20%20%20%20%20description%5Cn%20%20%20%20%20%20series%20%7B%5Cn%20%20%20%20%20%20%20%20label%5Cn%20%20%20%20%20%20%20%20points%20%7B%5Cn%20%20%20%20%20%20%20%20%20%20value%5Cn%20%20%20%20%20%20%20%20%7D%5Cn%20%20%20%20%20%20%7D%5Cn%20%20%20%20%7D%5Cn%20%20%7D%5Cn%7D%5Cn%22%2C%22variables%22%3A%22%22%7D)

**Comment by slimsag** [comment]
I deleted (restarts the container) the production codeinsights-db pod:

```
kubectl -n prod delete pod codeinsights-db-5f5977f74d-6gltl
```

And sure enough - now the entire DB is gone:

```
$ kubectl -n prod exec -it codeinsights-db-5f5977f74d-5hvss -- psql -U postgres
psql (12.5)
Type "help" for help.

postgres=# \d+
Did not find any relations.
```

So definitely something like https://github.com/sourcegraph/deploy-sourcegraph/pull/1936 causing the data loss issue here..

**Comment by slimsag** [comment]
![](https://i.imgur.com/ybWc6xa.gif)

This explains **everything**

When the DB gets lost (presumably because a new deployment of all services happens) it means that it looks like a brand new insight and starts over from scratch - EXCEPT for one thing: all of the jobs from the old insight are still in the queue, and will be executed in order. So if it got to where it was backfilling at 5 months ago, then now the queue of jobs will look like this:

- month 5
- month 6
- month 1
- month 2
- month 3
- month 4

Instead of being sequentially ordered (as it should) - THAT is what is giving us the crazy wonky charts:

![image](https://user-images.githubusercontent.com/3173176/113365865-5c28cb00-930c-11eb-8570-cb4a8031526a.png)

**Comment by slimsag** [comment]
OK - so why is data loss happening? Turns out this is specific to Sourcegraph.com. I did fix this, and it's fixed for all customers in the latest stable version of Sourcegraph - it just didn't roll out to sourcegraph.com (I guess our thing that merges github.com/sourcegraph/deploy-sourcegraph into github.com/sourcegraph/deploy-sourcegraph-dot-com is broken and it's just been out of sync for over a month now)

So that is one mystery solved (and good to know it won't affect customers trying this.)

**Comment by slimsag** [comment]
But what is going on with k8s.sgdev.org? it's not getting data loss, instead it's showing recorded `0` points (and the "data loss" fix has already rolled out there):

<img width="778" alt="image" src="https://user-images.githubusercontent.com/3173176/113366496-36042a80-930e-11eb-8055-5c693d6a9b8f.png">

**Comment by slimsag** [comment]
so on k8s.sgdev.org, it is tracking 3 backend insights (as expected):

```
postgres=# select count(distinct series_id) from series_points;
 count 
-------
     3
(1 row)
```

And we can see that the two we're looking at in the screenshot clearly have a tall number of data points (but we would expect around 40,000(repos) * 6(months) of data points):

```
postgres=# select count(*) from series_points GROUP BY series_id;
 count 
-------
 16719
  6093
    52
(3 rows)
```

We can see that these have barely scratched the surface of those 40k repos, by checking how many distinct repos those series are covering currently:

```
postgres=# select count(distinct repo_id) from series_points GROUP BY series_id;
 count 
-------
  5236
    16
  1946
(3 rows)
```

Very oddly, we can see that actually most of those data points are NOT zero:

```
postgres=# select count(*) from series_points where value>0 GROUP BY series_id;
 count 
-------
 16668
  6041
```

So what gives? Why is it displaying zeroes? Must be something to do with the backend query being ran - so [let's try that](https://docs.sourcegraph.com/dev/background-information/insights/backend#query-data-the-way-we-do-for-the-frontend-but-for-every-series-in-the-last-6mo) and sure enough we get zeroes. Hmm.

Playing with that query a bit, I see that if I drop the 6 month window I DO get data points. Are all the non-zero data points just older than six months? ...yup

```
postgres=# select count(*) from series_points where value>0 AND time > NOW() - INTERVAL '6 months' GROUP BY series_id;
 count 
-------
(0 rows)
```

So maybe those are from before (when the timeframe was "last 2 years" instead of "last 6 months") - did it record _any_ data points for the last 6mo? ...barely:

```
postgres=# select count(*) from series_points where time > NOW() - INTERVAL '6 months' GROUP BY series_id;
 count 
-------
    52
    52
    52
(3 rows)
```

exactly how many repos is this covering? just 16 apparently:

```
postgres=# select count(distinct repo_id) from series_points where time > NOW() - INTERVAL '6 months' GROUP BY series_id;
 count 
-------
    16
    16
    16
(3 rows)
```

so now we at least know the problem! it got to the 16th repo and then just... stopped? Why? Do the jobs exist? Let's check and..

```
$ kubectl -n dogfood-k8s exec -it pgsql-65bd6455d9-f64d9 -c pgsql -- psql -U sg
psql (12.6 (Debian 12.6-1.pgdg100+1))
Type "help" for help.

sg=# select state,count(*) from insights_query_runner_jobs GROUP BY state;
   state   |  count   
-----------+----------
 completed |  1891751
 failed    |      217
 queued    | 59517590
(3 rows)
```

WOW. So 1.8 million complete (presumably for those data points before 6mo) and 59 million search query jobs are just ready to go but absolutely nothing has been done to execute them it seems. Why the heck not?

**Comment by slimsag** [comment]
New question: is repo-updater on k8s.sgdev.org actually healthy? It's what runs the backend insights jobs, but it looks to be in super bad shape:

There are hundreds of these:

```
t=2021-04-02T00:43:06+0000 lvl=eror msg="Could not fetch account from authz provider" userID=29 authzProvider=https://bitbucket.sgdev.org/ error="no results returned by the Bitbucket Server API"
```

And these:

```
t=2021-04-02T00:43:06+0000 lvl=eror msg="Failed to sync permissions" type=2 id=31 err="fetch user permissions: no results returned by the Bitbucket Server API"
```

And these (**so code monitoring is also broken on k8s.sgdev.org right now?**):

```
t=2021-04-02T00:32:41+0000 lvl=warn msg="Marked record as errored" name=code_monitors_trigger_jobs_worker id=79413 err="Post: Post \"http://sourcegraph-frontend-internal/.internal/graphql?Search\": dial tcp 10.253.5.123:80: connect: connection refused"
```

And these:

```
2021/04/02 00:32:45 received error during background config update, err: unable to fetch new configuration: Post "http://sourcegraph-frontend-internal/.internal/configuration": dial tcp 10.253.5.123:80: connect: connection refused
```

And finally - our possible explanation for our problem - these:

```
t=2021-04-02T00:32:44+0000 lvl=warn msg="Marked record as errored" name=insights_query_runner_worker id=2149880 err="Post: Post \"http://sourcegraph-frontend-internal/.internal/graphql?InsightsSearch\": dial tcp 10.253.5.123:80: connect: connection refused"
```

So it seems like network activity on repo-updater on k8s.sgdev.org may be completely busted currently.

**Comment by slimsag** [comment]
Thread to try and resolve the k8s.sgdev.org repo-updater issues: https://sourcegraph.slack.com/archives/CMBA8F926/p1617324791088000

**Comment by slimsag** [comment]
# TL;DR

* repo-updater on k8s.sgdev.org is borked, code insights is working correctly there otherwise but because of that it only has very partial data it seems and is thus showing "0" data points (while really trying to indicate "I have literally not had a chance to do any work yet") - same issue has broken code monitoring it seems. https://sourcegraph.slack.com/archives/CMBA8F926/p1617324791088000
* All of the code insights issues on Sourcegraph.com (data resetting, and the really wonky looking charts) are explained by the fact that I fixed an issue a long time ago in https://github.com/sourcegraph/deploy-sourcegraph/pull/1936 but nothing rolled it out to sourcegraph.com - I don't know why, and haven't fixed this yet. Will need to ask distribution team. https://sourcegraph.slack.com/archives/CJX299FGE/p1617325370279700

**Comment by slimsag** [comment]
FYI, if we have time soon to tackle the frontend improvements described in https://github.com/sourcegraph/sourcegraph/issues/18708 then a majority of these issues would've been much more obvious:

* On k8s.sgdev.org, we could've had a "refreshing" icon that would show something like "56 million data points being queried..." as a hover tooltip on the "0 data points" chart.
* On sourcegraph.com, we could've had the same (still would've been confusion about the order of population, but would've made it a lot more clear that "it's doing stuff")

**Comment by Joelkw** [comment]
@slimsag thanks for all these comments, and noting the frontend issue could've helped here (cc @felixfbecker as well)

**Comment by felixfbecker** [comment]
@slimsag regarding you last comment, that would require backend changes first to expose those things through the GraphQL API, right? If so we're blocked on that

**Comment by slimsag** [comment]
Update:

* I will try to add something that can report from the backend the current status (just anything quick), I am running a bit behind so perhaps next week if I don't get to it tomorrow.
* I have checked on k8s.sgdev.org, that deployment is still _really_ broken and has remained so for the last 14d. I have pinged #dev-ops again. https://sourcegraph.slack.com/archives/CMBA8F926/p1618527985216400?thread_ts=1617324791.088000&cid=CMBA8F926
* I have sent a fix for sourcegraph.com that should roll out soon: https://github.com/sourcegraph/deploy-sourcegraph-dot-com/pull/5443

Once the fix for Sourcegraph.com goes out, I will wipe the work queue (which would otherwise have us "see the numbers come back in the wrong order again") and this issue should be fixed for sourcegraph.com entirely - and wouldn't have ever been an issue for customers AFAIK.

**Comment by slimsag** [comment]
I connected to prod DB using [these instructions](https://about.sourcegraph.com/handbook/engineering/deployments/postgresql#command-line-only-use-pgsql) (well, actually, I had issues with that and [got Dax to do it for me](https://sourcegraph.slack.com/archives/CJX299FGE/p1618530265078400)) and wiped the existing work queue using:

```
sg=> select COUNT(*) from insights_query_runner_jobs;
-> 39109486
```

```
sg=> TRUNCATE insights_query_runner_jobs;
-> deleted 39109486 rows
```

We should start to see the fmt usage chart make sense now. It should build out from right to left, with the leftmost point being inaccurate ("built up over time"). Currently its empty:

<img width="398" alt="image" src="https://user-images.githubusercontent.com/3173176/114953031-5d79ed80-9e0c-11eb-91ac-81a9d1443206.png">

**Comment by coury-clark** [comment]
Closing this issue. This issue was caused as described above as the out of order sequencing of points. Due to the duration a backfill can process, these charts can be very misleading until complete. Any new observations of anomalies will be opened and tracked in new and separate issues.
